### DISCUSSION TITLE
Help with dependency injection

### DISCUSSION THREAD

**Discussion by reelthymeos** [AvaloniaUI/Avalonia#17576] [Q&A] [answered]
Hi,

This is my first time working with dependency injection. I added Microsoft.Extensions.DependencyInjection to my Avalonia desktop project. I followed the quick Avalonia reference on DI here: https://docs.avaloniaui.net/docs/guides/implementation-guides/how-to-implement-dependency-injection
 
I have two questions

1. Can anyone point me to a more comprehensive tutorial or documentation source for DI?
2. Below is how I am approaching this app, please provide any pointers or suggestions

Basically the app is supposed to control a physical switching device. I created a class called 'InstrumentService' that represents a single instance of that switching device. But a user may have more than one switch device connected to their computer, so the app needs to be able to let the user select the device of interest and then it will display that device's info. So there may be multiple instances of 'InstrumentService' in existance, and the app should select the one that's needed. 

So I was thinking that I should have a List<InstrumentService> and that this should be 'registerted' so that the ViewModels will have access to it. 

I have  a MainViewModel which instantiates a bunch of sub-ViewModels as shown below. The sub-ViewModels need to also have access to the resgistered service (List of InstrumentService) 

Below is my Main View Model:

```
public partial class MainViewModel
{
	private List<InstrumentService> instrumentList = new List<InstrumentService>();

	public ChildViewModel_1 childViewModel_1 { get; private set; }
	public ChildViewModel_2 childViewModel_2 { get; private set; }
	public ChildViewModel_3 childViewModel_3 { get; private set; }
	
	public MainViewModel()
	{
		childViewModel_1 = new ChildViewModel_1;
		childViewModel_2 = new ChildViewModel_2;
		childViewModel_3 = new ChildViewModel_3;		
	}

}
```

I created this class per the Avalonia document:

```
public static class ServiceCollectionExtensions
{
		public static void AddCommonServices(this IServiceCollection collection)
		{
			collection.AddTransient<InstrumentService>();
			collection.AddTransient<MainViewModel>();
		}
}
```

Then I did this in App.axaml.cs:

```
public override void OnFrameworkInitializationCompleted()
    {
        // Line below is needed to remove Avalonia data validation.
        // Without this line you will get duplicate validations from both Avalonia and CT
        BindingPlugins.DataValidators.RemoveAt(0);

        // Register all the services needed for the application to run
        var collection = new ServiceCollection();
        collection.AddCommonServices();

        // Creates a ServiceProvider containing services from the provided IServiceCollection
        var services = collection.BuildServiceProvider();
        var vm = services.GetRequiredService<MainViewModel>();

        if (ApplicationLifetime is IClassicDesktopStyleApplicationLifetime desktop)
        {
            desktop.MainWindow = new MainWindow
            {
               // DataContext = new MainViewModel()
                  DataContext = vm
            };
        }
}
```

But how can I access this from the various sub-ViewModels? i.e. how can they access the list of InstrumentService?

Thanks

**Comment by thevortexcloud** [comment, marked as answer]
> Can anyone point me to a more comprehensive tutorial or documentation source for DI?

If you want a really detailed guide, Microsoft has their own. But it's mainly orientated around using it with ASP Core, so not all of it applies to using it with something like Avalonia.

https://learn.microsoft.com/en-us/dotnet/core/extensions/dependency-injection-basics
https://learn.microsoft.com/en-us/dotnet/core/extensions/dependency-injection
https://learn.microsoft.com/en-us/aspnet/core/fundamentals/dependency-injection?view=aspnetcore-9.0

> So I was thinking that I should have a List and that this should be 'registerted' so that the ViewModels will have access to it.

The MSDI approach is actually to just register multiple implementations of the same service, then change your constructor to accept an IEnumerable of them. That way it will just pass all of them in for you. If you don't use the IEnumerable, it will just pass the first one in.

> But how can I access this from the various sub-ViewModels? i.e. how can they access the list of InstrumentService?

Register your child view models as services, and then inject them into your main view model. The DI system builds up a whole dependency tree and works out what services need access to other services. The result is you don't really have to think about passing services around, you just need to make sure they are registered.

If you have never used DI before, it can take awhile to understand how to use it properly especially since it's kind of backwards from the normal C# way of doing things.

**Comment by fallenleavesguy** [comment]
https://learn.microsoft.com/en-us/dotnet/communitytoolkit/mvvm/ioc

**Reply by thevortexcloud** [reply]
I am not sure why you linked to the same guide linked in the OP?

**Reply by fallenleavesguy** [reply]
> I am not sure why you linked to the same guide linked in the OP?
I paste wrong. Now it links to right place
